### ISSUE TITLE
Memory leak in ``timedesc`` node order.

### ISSUE THREAD

**Issue by Bai-fengting** [tskit-dev/tskit#2423]
Hello,
When I running the flowing code, I found the memroy increased very quickly (until more that 100G and I killed it).
It is wired, because in each iteration there are only one tree in process. Why the memory inceased along with the iteration.
Furthermor, when I add ts.nbytes, the memory  no icreased so quickly. But the speed became slower obviously.
How I can fix that?
```python
for i in range(0, 10000):
#    ts.nbytes
    tree_i = ts.at_index(i)
    num_lineages = {}
    count = 1
    for node in tree_i.nodes(order="timedesc"):
        if tree_i.is_sample(node):
            num_lineages[node] = tree_i.num_samples()
        else:
            num_lineages[node] = count
            count += tree_i.num_children(node) - 1
```

**Comment by jeromekelleher** [comment]
Hi @Bai-fengting, :wave:

You're right, this code chunk does have a serious memory leak - I'm not sure why yet, but I have reproduced it. Luckily, there's a much more efficient way to do what you want which doesn't have a memory leak:
```python
for tree_i in ts.trees():
    if tree_i.index == 10000:
        break
    num_lineages = {}
    count = 1
    for node in tree_i.nodes(order="timedesc"):
        if tree_i.is_sample(node):
            num_lineages[node] = tree_i.num_samples()
        else:
            num_lineages[node] = count
        count += tree_i.num_children(node) - 1
```

Iterating over the trees like this is **much** more efficient than using ``ts.at()`` repeatedly.

**Comment by jeromekelleher** [comment]
Digging in a bit, this is something to do with using the ``timedesc`` order, which involves using ``self.tree_sequence.tables``.  Not sure why it's leaking memory so much, but there must be some reference counting going wrong somewhere.

This is a bug, we should try and fix this specific problem with ``timedesc`` anyway even if we're not getting at the underlying problem with reference counting in ``tree_sequence.tables``.

**Comment by jeromekelleher** [comment]
@Bai-fengting, I'm not so sure that my solution above is actually resolving things - can you try it out and let me know if it fixes your memory leak problem please?

**Comment by Bai-fengting** [comment]
> @Bai-fengting, I'm not so sure that my solution above is actually resolving things - can you try it out and let me know if it fixes your memory leak problem please?

Thanks @jeromekelleher , I tried the above code. But the problem not resolved, the memory still increased quickly.
As you mentioned, it caused by `timedesc` order. After I remove `order="timedesc"`, the memory not leak.
In fact, I need the count the number of lineages above a specific node in each tree. I used the code mentioned in this issue:
https://github.com/tskit-dev/tskit/issues/2190#issuecomment-1090889782

Is there other way to do that?

**Comment by jeromekelleher** [comment]
Thanks @Bai-fengting - I'm working on a fix for the leak over in #2417 

As it happens, I've also got an efficient implementation of num_lineages going over in #2422. Hopefully both of these will be released in a week or so.

In the meantime, your code above should work in the same way if you use ``order="postorder"`` instead of ``timedesc`` (and be much more efficient).
